docs/spec.md

```diff
@@ -2288,23 +2288,59 @@ Computes dot products between slices of `lhs` and slices of `rhs` and produces a
+* `If lhs uses non-quantized tensor and rhs uses quantized tensor:`
+
+  ```python
+   for batch_size in range(0, dim(lhs, lhs_batching_dimensions[0])):
+    rmin = lhs[batch_size, :].min()
+    rmax = lhs[batch_size, :].max()
+    scale = (rmax - rmin) / (storage_max(rhs) - storage_min(rhs))
+    zero_point = (rmax * storage_min(rhs) - rmin * storage_max(rhs)) / (rmax - rmin)
+    effective_lhs[batch_size, :] = uniform.quantize(lhs[batch_size, :],
+            storage_type=storage_type(rhs),
+            storage_min=storage_min(rhs),
+            storage_max=storage_max(rhs),
+            quantization_dimension=lhs_batching_dimensions[0],
+            scale=scale, zero_point=zero_point)
+  ```
+
+* `If lhs uses quantized tensor or rhs uses non-quantized tensor:`
+  * `effective_lhs = lhs`.
+* `lhs_result_dimensions = [d for d in axes(effective_lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
 * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
 * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
   where `size(result_batching_index) = size(lhs_batching_dimensions)`,
   `size(result_lhs_index) = size(lhs_result_dimensions)` and
   `size(result_rhs_index) = size(rhs_result_dimensions)`.
-* `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
-* `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ..., :])`.
-* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
+* `transposed_lhs = transpose(effective_lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
+* `transposed_lhs_slice = slice(transposed_lhs, result_batching_index + result_lhs_index + [:, ..., :])`.
+* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(effective_lhs, lhs_contracting_dimensions))`.
 * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
-* `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ..., :])`.
+* `transposed_rhs_slice = slice(transposed_rhs, result_batching_index + result_rhs_index + [:, ..., :])`.
 * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
-* `dot_product = reduce(
-  inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
-  init_values=[0],
-  dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
-  body=lambda x, y: add(x, y))`.
+* For `is_non_quantized_tensor(lhs) and is_non_quantized_tensor(rhs)`:
+  * `dot_product = reduce(
+      inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
+      init_values=[0],
+      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+      body=lambda x, y: add(x, y))`.
+* For `is_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
+  * `integer_dot_product = reduce(
```

Comment:
Does the specification define the type or range for integer_dot_product? For example, for 8-bit quantized data I would expect integer_dot_product to have 32-bit integer range. For 16-bit quantized data I would expect a larger range such as 48-bit integer.

Reply:
That a good point and you are right that the specifics of accumulation range is not specified in the specification document for ops like dot_general or convolution. IMO, the problem is not limited to quantized types but non-quantized types as well, which allows us to isolate it from the current PR. We have an open issue to address that https://github.com/openxla/stablehlo/issues/755. (As mention in a [comment](https://github.com/openxla/stablehlo/issues/1439#issuecomment-1527815787),  precision_config has ties with the accumulation range).

Let me know if you have any feedback on the above arrangement.

Reply:
Thanks @sdasgup3 for the link to [#755](https://github.com/openxla/stablehlo/issues/755) and related comments. I agree these issues are related. The precision_config described in [#755](https://github.com/openxla/stablehlo/issues/755) is mainly talking about accuracy of floating-point accumulations in terms of DEFAULT, HIGH, HIGHEST. For an integer sum, I think the main issue is likely to be accumulator overflow rather than accuracy. For this the operation should describe the minimum accumulator width supported.

There are some operations where different accumulator widths may be useful to select. For example, for quantized 16-bit by 8-bit dot product operations, both 32-bit and 48-bit accumulation range may be useful depending on the usage. This could be selected by something like DEFAULT/HIGH but it would be range rather than accuracy.

Are there plans to expand [#755](https://github.com/openxla/stablehlo/issues/755) to cover quantized accumulator widths as part of the precision definition?

Reply:
Thanks @dominicsymes for explaining. I agree with you. 

>> the operation should describe the minimum accumulator width supported.

As you said,  the `precision_config`  covers part of the picture (the precision at which the computation should happens). In general, for dot_general or convolution, the accumulation should be done in a range determined by at least  the element-type of the output tensor type (which [preferred_element_type](https://github.com/openxla/stablehlo/issues/600) can override).

I should say that the combination of #755 and #600 should help bring clarity into this.

Reply:
Thanks @sdasgup3  - I agree some combination of #755 and #600 should work (but output type is not sufficient on its own for accumulation range, as it is after scaling). I’ve added a comment in [#755](https://github.com/openxla/stablehlo/issues/755) to note the integer use case as well.

Reply:
Thanks @dominicsymes for adding the information. I will explore more on this as  part of the issue.